File: day0/K/Wonderseabreeze/runners.py
import numpy as np

result = {}
patterns = [
	np.array([[1, 1]]),
	np.array([[1, 1, 1]]),
	np.array([[1, 1], [0, 1]]),
	np.array([[1, 1, 1, 1]]),
	np.array([[1, 1], [1, 1]]),
	np.array([[1, 1, 0], [0, 1, 1]]),
	np.array([[1, 1, 1], [0, 1, 0]]),
	np.array([[1, 1, 1], [1, 0, 0]])
]
pattern = np.zeros((4, 4))
LENGTH = 5

get_result = lambda step: 1 if step % 3 == 0 else -1

def check(state, player):
	global pattern
	# for rot in range(4):
	for i in range(LENGTH + 1 - pattern.shape[0]):
		for j in range(LENGTH + 1 - pattern.shape[1]):
			for x in range(pattern.shape[0]):
				for y in range(pattern.shape[1]):
					if pattern[x, y] and state[i + x, j + y] != player:
						break
				else:
					continue
				break
			else:
				# if pattern.sum() == 3:
					# print(state, player)
				return True
	# pattern = np.rot90(pattern)
	return False

def compress(state):
	return "".join(map(str, state.flatten()))

def Alice(step, state, comp):
	# comp = compress(state)
	if comp in result:
		return result[comp]
	best = -2
	for x in range(LENGTH):
		for y in range(LENGTH):
			if state[x, y] == 0:
				state[x, y] = 1
				if check(state, 1):
					now = get_result(step)
				else:
					now = Bob(step + 1, state, comp + 2 ** (x * 5 + y))
				state[x, y] = 0
				if now == 1:
					result[comp] = 1
					# print(pattern, state, f"({x}, {y})")
					return 1
				if now > best:
					best = now
	result[comp] = best
	return best

def Bob(step, state, comp):
	# comp = compress(state)
	if comp in result:
		return result[comp]
	best = 2
	for x in range(LENGTH):
		for y in range(LENGTH):
			if state[x, y] == 0:
				# state[x, y] = 2
				# if check(state, 2):
				state[x, y] = 1
				if check(state, 1):
					now = -get_result(step)
				else:
					now = Alice(step + 1, state, comp + 2 ** (x * 5 + y))
				state[x, y] = 0
				if now == -1:
					result[comp] = -1
					return -1
				if now < best:
					best = now
	result[comp] = best
	return best



for pattern in patterns:
	state = np.zeros((LENGTH, LENGTH), dtype = np.int32)
	print(pattern, Alice(1, state, 0), flush = True)
	result = {}

File: day0/K/Wonderseabreeze/runners.cpp
#include <cstdio>
#include <cstring>
using namespace std;
#define Result(_x) ((_x) % 3 == 0 ? 1 : -1)
#define SIZE 5
int state[SIZE][SIZE], pattern[SIZE][SIZE], res[33660000], ph, pw;
int check(int x, int y){
	int marken = 0;
	for (int i = 0; i < ph && i <= x; ++i) {
		for (int j = 0; j < pw && j <= y; ++j) {
			if (pattern[i][j] && x - i + ph <= SIZE && y - j + pw <= SIZE) {
				marken = 1;
				for (int dx = 0; dx < ph && marken; ++dx) {
					for (int dy = 0; dy < pw; ++dy) {
						if (pattern[dx][dy] && state[x - i + dx][y - j + dy] == 0) {
							marken = 0;
							break;
						}
					}
				}
				if (marken) return 1;
			}
		}
	}
	/*
	for (int i = 0; i + ph <= SIZE; ++i) {
		for (int j = 0; j + pw <= SIZE; ++j) {
			marken = 1;
			for (int dx = 0; dx < ph && marken; ++dx) {
				for (int dy = 0; dy < pw; ++dy) {
					if (pattern[dx][dy] && state[i + dx][j + dy] == 0) {
						marken = 0;
						break;
					}
				}
			}
			if (marken) return 1;
		}
	}
	*/
	return 0;
}
int Bob(int step, int s);
int Alice(int step, int s) {
	if (res[s]) return res[s];
	int best = -2, now;
	for (int i = 0; i < SIZE; ++i) {
		for (int j = 0; j < SIZE; ++j) {
			if (state[i][j] == 0) {
				state[i][j] = 1;
				if (check(i, j)) {
					now = Result(step);
				}
				else {
					now = Bob(step + 1, s | 1 << (i * 5 + j));
				}
				state[i][j] = 0;
				if (now > best) {
					best = now;
					if (best == 1) return res[s] = 1;
				}
			}
		}
	}
	return res[s] = best;
}
int Bob(int step, int s) {
	if (res[s]) return res[s];
	int best = 2, now;
	for (int i = 0; i < SIZE; ++i) {
		for (int j = 0; j < SIZE; ++j) {
			if (state[i][j] == 0) {
				state[i][j] = 1;
				if (check(i, j)) {
					now = -Result(step);
				}
				else {
					now = Alice(step + 1, s | 1 << (i * 5 + j));
				}
				state[i][j] = 0;
				if (now < best) {
					best = now;
					if (best == -1) return res[s] = -1;
				}
			}
		}
	}
	return res[s] = best;
}
int main(){
	ph = 1, pw = 2;
	pattern[0][0] = pattern[0][1] = 1;
	printf("++: %d\n", Alice(1, 0));
	
	memset(res, 0, sizeof(res));
	
	ph = 1, pw = 3;
	pattern[0][0] = pattern[0][1] = pattern[0][2] = 1;
	printf("+++: %d\n", Alice(1, 0));
	
	memset(res, 0, sizeof(res));
	
	ph = 2, pw = 2;
	pattern[0][0] = pattern[0][1] = pattern[1][0] = 1;
	pattern[1][1] = 0;
	printf("++\n+ : %d\n", Alice(1, 0));
	
	memset(res, 0, sizeof(res));
	
	ph = 1, pw = 4;
	pattern[0][0] = pattern[0][1] = pattern[0][2] = pattern[0][3] = 1;
	printf("++++: %d\n", Alice(1, 0));
	
	memset(res, 0, sizeof(res));
	
	ph = 2, pw = 2;
	pattern[0][0] = pattern[0][1] = pattern[1][0] = pattern[1][1] = 1;
	printf("++\n++: %d\n", Alice(1, 0));
	
	memset(res, 0, sizeof(res));
	
	ph = 2, pw = 3;
	pattern[0][0] = pattern[0][1] = pattern[0][2] = pattern[1][0] = 1;
	pattern[1][1] = pattern[1][2] = 0;
	printf("+++\n+  : %d\n", Alice(1, 0));
	
	memset(res, 0, sizeof(res));
	
	ph = 2, pw = 3;
	pattern[0][0] = pattern[0][1] = pattern[0][2] = pattern[1][1] = 1;
	pattern[1][0] = pattern[1][2] = 0;
	printf("+++\n + : %d\n", Alice(1, 0));
	
	memset(res, 0, sizeof(res));
	
	ph = 2, pw = 3;
	pattern[0][0] = pattern[0][1] = pattern[0][2] = pattern[1][2] = 1;
	pattern[1][0] = pattern[1][1] = 0;
	printf("+++\n  +: %d\n", Alice(1, 0));
	
	memset(res, 0, sizeof(res));
	
	return 0;
}